From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Initial Source <auto@mated.null>
Date: Tue, 6 Jul 2021 11:36:49 -0500
Subject: [PATCH] Initial


diff --git a/build.gradle.kts b/build.gradle.kts
index e8bd6b61b1d9cbec3fe1b9003840ec6224efe036..cbd7a37dd8f618a9b9a605d13d61866dd9930b90 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -3,7 +3,6 @@ import com.github.jengelman.gradle.plugins.shadow.transformers.Transformer
 import io.papermc.paperweight.tasks.BaseTask
 import io.papermc.paperweight.tasks.GenerateReobfMappings
 import io.papermc.paperweight.tasks.PatchMappings
-import io.papermc.paperweight.util.Constants
 import io.papermc.paperweight.util.Git
 import io.papermc.paperweight.util.cache
 import io.papermc.paperweight.util.defaultOutput
@@ -112,22 +111,6 @@ publishing {
     }
 }
 
-val generateReobfMappings = rootProject.tasks.named<GenerateReobfMappings>("generateReobfMappings")
-
-val patchReobfMappings by tasks.registering<PatchMappings> {
-    inputMappings.set(generateReobfMappings.flatMap { it.reobfMappings })
-    patch.set(rootProject.layout.cache.resolve("paperweight/upstreams/paper/build-data/reobf-mappings-patch.tiny"))
-
-    fromNamespace.set(Constants.DEOBF_NAMESPACE)
-    toNamespace.set(Constants.SPIGOT_NAMESPACE)
-
-    outputMappings.set(layout.cache.resolve("paperweight/mappings/reobf-patched.tiny"))
-}
-
-tasks.reobfJar {
-    mappingsFile.set(patchReobfMappings.flatMap { it.outputMappings })
-}
-
 val generatePom = tasks.named<GenerateMavenPom>("generatePomFileForMavenPublication")
 
 tasks.shadowJar {
