From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Paul <paul@technove.co>
Date: Wed, 30 Jun 2021 13:50:29 -0500
Subject: [PATCH] idk


diff --git a/build.gradle.kts b/build.gradle.kts
index f766183883f009368b79e1201553c184a373e2cf..76fbcadd5637bccda2ee44d834128f3dd462eb6a 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -162,7 +162,7 @@ abstract class IncludeMappings : BaseTask() {
 }
 
 val includeMappings = tasks.register<IncludeMappings>("includeMappings") {
-    inputJar.set(tasks.fixJarForReobf.flatMap { it.outputJar })
+    inputJar.set(tasks.shadowJar.flatMap { it.archiveFile }) // Airplane
     mappings.set(tasks.reobfJar.flatMap { it.mappingsFile })
 }
 
diff --git a/src/main/java/net/minecraft/core/BlockPos.java b/src/main/java/net/minecraft/core/BlockPos.java
index b01d7da333bac7820e42b6f645634a15ef88ae4f..43e5c3fde5d45ef975a8ddbe4a2758ea230dddd4 100644
--- a/src/main/java/net/minecraft/core/BlockPos.java
+++ b/src/main/java/net/minecraft/core/BlockPos.java
@@ -478,9 +478,9 @@ public class BlockPos extends Vec3i {
         }
 
         public BlockPos.MutableBlockPos set(int x, int y, int z) {
-            this.x = x; // Tuinity - force inline
-            this.y = y; // Tuinity - force inline
-            this.z = z; // Tuinity - force inline
+            ((Vec3i) this).x = x; // Tuinity - force inline
+            ((Vec3i) this).y = y; // Tuinity - force inline
+            ((Vec3i) this).z = z; // Tuinity - force inline
             return this;
         }
 
@@ -544,19 +544,19 @@ public class BlockPos extends Vec3i {
         // Paper start - comment out useless overrides @Override - TODO figure out why this is suddenly important to keep
         @Override
         public BlockPos.MutableBlockPos setX(int i) {
-            this.x = i; // Tuinity
+            ((Vec3i) this).x = i; // Tuinity
             return this;
         }
 
         @Override
         public BlockPos.MutableBlockPos setY(int i) {
-            this.y = i; // Tuinity
+            ((Vec3i) this).y = i; // Tuinity
             return this;
         }
 
         @Override
         public BlockPos.MutableBlockPos setZ(int i) {
-            this.z = i; // Tuinity
+            ((Vec3i) this).z = i; // Tuinity
             return this;
         }
         // Paper end
