From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Paul <paul@technove.co>
Date: Wed, 30 Jun 2021 13:50:29 -0500
Subject: [PATCH] idk


diff --git a/build.gradle.kts b/build.gradle.kts
index 0837e11261af88aa86d9f0d1f1167f2fe292db90..ac83dfed15027b8f6eb1ef4115332bc04479297a 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -185,7 +185,7 @@ abstract class IncludeMappings : BaseTask() {
 }
 
 val includeMappings = tasks.register<IncludeMappings>("includeMappings") {
-    inputJar.set(tasks.fixJarForReobf.flatMap { it.outputJar })
+    inputJar.set(tasks.shadowJar.flatMap { it.archiveFile }) // Airplane
     mappings.set(tasks.reobfJar.flatMap { it.mappingsFile })
 }
 
diff --git a/src/main/java/net/minecraft/core/BlockPos.java b/src/main/java/net/minecraft/core/BlockPos.java
index 30a4b80eacf76ad7f0a48b79bcf01b752c0af48a..177134c0f232e04cb081c3182340f764fbaf4b8e 100644
--- a/src/main/java/net/minecraft/core/BlockPos.java
+++ b/src/main/java/net/minecraft/core/BlockPos.java
@@ -484,9 +484,9 @@ public class BlockPos extends Vec3i {
         }
 
         public BlockPos.MutableBlockPos set(int x, int y, int z) {
-            this.x = x; // Tuinity - force inline
-            this.y = y; // Tuinity - force inline
-            this.z = z; // Tuinity - force inline
+            ((Vec3i) this).x = x; // Tuinity - force inline
+            ((Vec3i) this).y = y; // Tuinity - force inline
+            ((Vec3i) this).z = z; // Tuinity - force inline
             return this;
         }
 
@@ -550,19 +550,19 @@ public class BlockPos extends Vec3i {
         // Paper start - comment out useless overrides @Override - TODO figure out why this is suddenly important to keep
         @Override
         public BlockPos.MutableBlockPos setX(int i) {
-            this.x = i; // Tuinity
+            ((Vec3i) this).x = i; // Tuinity
             return this;
         }
 
         @Override
         public BlockPos.MutableBlockPos setY(int i) {
-            this.y = i; // Tuinity
+            ((Vec3i) this).y = i; // Tuinity
             return this;
         }
 
         @Override
         public BlockPos.MutableBlockPos setZ(int i) {
-            this.z = i; // Tuinity
+            ((Vec3i) this).z = i; // Tuinity
             return this;
         }
         // Paper end
