From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Initial Source <auto@mated.null>
Date: Mon, 21 Jun 2021 15:25:32 -0500
Subject: [PATCH] Initial


diff --git a/build.gradle.kts b/build.gradle.kts
index a22cedb52fa792011ba83480bfc5bed8ac59580b..8fbd6c3e160f76e9aa98330e456fc9e28cdfdf29 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -3,7 +3,6 @@ import com.github.jengelman.gradle.plugins.shadow.transformers.Transformer
 import io.papermc.paperweight.tasks.BaseTask
 import io.papermc.paperweight.tasks.GenerateReobfMappings
 import io.papermc.paperweight.tasks.PatchMappings
-import io.papermc.paperweight.util.Constants
 import io.papermc.paperweight.util.Git
 import io.papermc.paperweight.util.cache
 import io.papermc.paperweight.util.defaultOutput
@@ -113,14 +112,15 @@ publishing {
     }
 }
 
+
 val generateReobfMappings = rootProject.tasks.named<GenerateReobfMappings>("generateReobfMappings")
 
 val patchReobfMappings by tasks.registering<PatchMappings> {
     inputMappings.set(generateReobfMappings.flatMap { it.reobfMappings })
     patch.set(rootProject.layout.cache.resolve("paperweight/upstreams/paper/build-data/reobf-mappings-patch.tiny"))
 
-    fromNamespace.set(Constants.DEOBF_NAMESPACE)
-    toNamespace.set(Constants.SPIGOT_NAMESPACE)
+    fromNamespace.set(io.papermc.paperweight.util.Constants.DEOBF_NAMESPACE)
+    toNamespace.set(io.papermc.paperweight.util.Constants.SPIGOT_NAMESPACE)
 
     outputMappings.set(layout.cache.resolve("paperweight/mappings/reobf-patched.tiny"))
 }
