From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Paul Sauve <paul@technove.co>
Date: Sat, 31 Oct 2020 19:03:25 -0500
Subject: [PATCH] Queue lighting update only once


diff --git a/src/main/java/net/minecraft/server/ChunkProviderServer.java b/src/main/java/net/minecraft/server/ChunkProviderServer.java
index 12d9b73ccc2f4406957932397746cac7902d650e..fc9919dd01cad02486af293f9b0c692046c7eea1 100644
--- a/src/main/java/net/minecraft/server/ChunkProviderServer.java
+++ b/src/main/java/net/minecraft/server/ChunkProviderServer.java
@@ -1005,6 +1005,7 @@ public class ChunkProviderServer extends IChunkProvider {
             // Paper - moved up
             // Tuinity start - optimise chunk tick iteration
             com.tuinity.tuinity.util.maplist.IteratorSafeOrderedReferenceSet.Iterator<Chunk> iterator = this.entityTickingChunks.iterator();
+            boolean updateLighting = false; // AirplaneL
             try {
             while (iterator.hasNext()) {
                 Chunk chunk = iterator.next();
@@ -1029,7 +1030,7 @@ public class ChunkProviderServer extends IChunkProvider {
                             }
 
                             this.world.timings.chunkTicks.startTiming(); // Spigot // Paper
-                            this.world.a(chunk, k);
+                            if (this.world.abool(chunk, k)) updateLighting = true; // AirplaneL
                             this.world.timings.chunkTicks.stopTiming(); // Spigot // Paper
                             MinecraftServer.getServer().executeMidTickTasks(); // Tuinity - exec chunk tasks during world tick
                         }
@@ -1039,6 +1040,7 @@ public class ChunkProviderServer extends IChunkProvider {
             } finally {
                 iterator.finishedIterating();
             }
+            if (updateLighting) this.getLightEngine().queueUpdate(); // AirplaneL
             // Tuinity end - optimise chunk tick iteration
             this.world.getMethodProfiler().enter("customSpawners");
             if (flag1) {
diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
index 8d8385e3022dd84e2b92daeffa190dbda60c705b..18aec64c92ef1cf24785244fe511d9beb609bf6b 100644
--- a/src/main/java/net/minecraft/server/WorldServer.java
+++ b/src/main/java/net/minecraft/server/WorldServer.java
@@ -971,7 +971,10 @@ public class WorldServer extends World implements GeneratorAccessSeed {
     private final com.destroystokyo.paper.util.math.ThreadUnsafeRandom randomTickRandom = new com.destroystokyo.paper.util.math.ThreadUnsafeRandom();
     // Paper end
 
-    public void a(Chunk chunk, int i) { final int randomTickSpeed = i; // Paper
+    // AirplaneL start - create version of chunk tick that returns a bool for updating lighting
+    public void a(Chunk chunk, int i) { this.abool(chunk, i); }
+    public boolean abool(Chunk chunk, int i) { final int randomTickSpeed = i; // Paper
+        // AirplaneL end
         ChunkCoordIntPair chunkcoordintpair = chunk.getPos();
         boolean flag = this.isRaining();
         int j = chunkcoordintpair.d();
@@ -1074,9 +1077,13 @@ public class WorldServer extends World implements GeneratorAccessSeed {
             }
             gameprofilerfiller.exit();
             timings.chunkTicksBlocks.stopTiming(); // Paper
-            getChunkProvider().getLightEngine().queueUpdate(); // Paper
+            // AirplaneL start
+            //getChunkProvider().getLightEngine().queueUpdate(); // Paper
+            return true;
+            // AirplaneL end
             // Paper end
         }
+        return false; // AirplaneL
     }
 
     protected BlockPosition a(BlockPosition blockposition) {
