From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Paul <paul@technove.co>
Date: Mon, 6 Sep 2021 12:56:49 -0500
Subject: [PATCH] Purpur Tweaks


diff --git a/src/main/java/gg/airplane/compat/ServerConfigurations.java b/src/main/java/gg/airplane/compat/ServerConfigurations.java
index f9a71ff3edd7e7b6cda680e5a156373b5aa813c2..da9880b0e5e1e8aed5c50d27e878787bb15affb3 100644
--- a/src/main/java/gg/airplane/compat/ServerConfigurations.java
+++ b/src/main/java/gg/airplane/compat/ServerConfigurations.java
@@ -24,7 +24,8 @@ public class ServerConfigurations {
       "bukkit.yml",
       "spigot.yml",
       "paper.yml",
-      "airplane.yml"
+      "airplane.yml",
+      "purpur.yml"
     };
 
     public static Map<String, String> getCleanCopies() throws IOException {
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 3c471f406d742e0eb8ac8c35dc807b109d9b185c..6066e9f9682b33dafab4ae8e4b23e3b446023b5c 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1737,7 +1737,10 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 
     @DontObfuscate
     public String getServerModName() {
-        return net.pl3x.purpur.PurpurConfig.serverModName; // Purpur - Purpur > // Airplane > // Paper - Paper > // Spigot - Spigot > // CraftBukkit - cb > vanilla!
+        // Airplane start
+        String purpurName = net.pl3x.purpur.PurpurConfig.serverModName;
+        return purpurName.equals("Purpur") ? "AirplanePurpur" : purpurName; // Purpur - Purpur > // Airplane > //Paper - Paper > // Spigot - Spigot > // CraftBukkit - cb > vanilla!
+        // Airplane end
     }
 
     public SystemReport fillSystemReport(SystemReport details) {
