From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jacob <jacob@stellardev.org>
Date: Wed, 19 Jan 2022 17:09:01 -0700
Subject: [PATCH] Toggle Tick Inactive Goals


diff --git a/src/main/java/com.nftworlds/NFTWorldsConfig.java b/src/main/java/com.nftworlds/NFTWorldsConfig.java
index 2796c77d6f8e91b8a62c4a57de33d38dce4c449d..bf3511978652f00a93be2bbb159a74f0baf9b50b 100644
--- a/src/main/java/com.nftworlds/NFTWorldsConfig.java
+++ b/src/main/java/com.nftworlds/NFTWorldsConfig.java
@@ -200,4 +200,7 @@ public class NFTWorldsConfig {
     public static boolean asyncEntities = true;
     private static void asyncEntities() { asyncEntities = getBoolean("async-entities.enabled", true); }
 
+    public static boolean tickInactiveGoals = false;
+    private static void tickInactiveGoals() { tickInactiveGoals = getBoolean("tick-inactive-goals", false); }
+
 }
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index b247d9dc2b2f5b4a6e87ca35e2a08f9be4e7a7b4..a326b96e53a7f06ecfe63f0bdc6e3d27e02573e0 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -9,6 +9,8 @@ import java.util.Optional;
 import java.util.Random;
 import java.util.UUID;
 import javax.annotation.Nullable;
+
+import com.nftworlds.NFTWorldsConfig;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.NonNullList;
 import net.minecraft.core.particles.ParticleTypes;
@@ -207,7 +209,7 @@ public abstract class Mob extends LivingEntity {
     @Override
     public void inactiveTick() {
         super.inactiveTick();
-        if (this.goalSelector.inactiveTick(this.activatedPriority, true)) { // Airplane - pass activated priroity
+        if (this.goalSelector.inactiveTick(this.activatedPriority, true) && NFTWorldsConfig.tickInactiveGoals) { // Airplane - pass activated priroity
             this.goalSelector.tick();
         }
         if (this.targetSelector.inactiveTick(this.activatedPriority, true)) { // Airplane - pass activated priority
